coap/messages.go

```diff
@@ -0,0 +1,154 @@
+package adapter
+
+import (
+	"bytes"
+	"log"
+	"net"
+
+	mux "github.com/dereulenspiegel/coap-mux"
+	coap "github.com/dustin/go-coap"
+	"github.com/mainflux/mainflux/coap/nats"
+	"github.com/mainflux/mainflux/writer"
+)
+
+// sendMessage sends the message to NATS
+func (ca *CoAPAdapter) sendMessage(l *net.UDPConn, a *net.UDPAddr, m *coap.Message) *coap.Message {
+	log.Printf("Got message in sendMessage: path=%q: %#v from %v", m.Path(), m, a)
+	var res *coap.Message = nil
+	if m.IsConfirmable() {
+		res = &coap.Message{
+			Type:      coap.Acknowledgement,
+			Code:      coap.Content,
+			MessageID: m.MessageID,
+			Token:     m.Token,
+			Payload:   []byte(""),
+		}
+		res.SetOption(coap.ContentFormat, coap.AppJSON)
+	}
+
+	if len(m.Payload) == 0 {
+		if m.IsConfirmable() {
+			res.Payload = []byte("{\"res\": \"Error: msg len can not be 0\"}")
+		}
+		return res
+	}
+
+	// Channel ID
+	cid := mux.Var(m, "channel_id")
+
+	// Publish message via NATS
+	n := writer.RawMessage{}
+	n.Channel = cid
+	n.Publisher = ""
+	n.Protocol = "coap"
+	n.Payload = m.Payload
+
+	if err := nats.Send(n); err != nil {
+		if m.IsConfirmable() {
+			res.Payload = []byte("{\"error\": \"cannot publish\"}")
```

Comment:
What is payload? Is this something you return to the clients? If so, drop it, since no other protocols perform the same action, and stuff like `{"error":"cannot publish"}`, or `{"res":"sent"}` provide absolutely no more information to the clients than bare response status.